graphql-server/pkg/dataloaders/proposal.go

```diff
@@ -60,9 +79,101 @@ func NewProposalDataloader(proposalQuery queries.IProposalQuery) ProposalDataloa
 		Wait:     DefaultWait,
 	})
 
+	proposalVoteLoader := godataloader.NewDataLoader(godataloader.DataLoaderConfig[ProposalVoteKey, *models.ProposalVote]{
+		MaxBatch: DefaultMaxBatch,
+		Wait:     DefaultWait,
+		Fetch: func(keys []ProposalVoteKey) ([]*models.ProposalVote, []error) {
+			// use maps to check if proposal id or address have been added to array yet
+			proposalIDsMap := make(map[int]struct{})
+			addressesMap := make(map[string]struct{})
+
+			proposalIDs := []int{}
+			addresses := []string{}
+			for _, key := range keys {
+				if _, exists := proposalIDsMap[key.ProposalID]; !exists {
+					proposalIDsMap[key.ProposalID] = struct{}{}
+					proposalIDs = append(proposalIDs, key.ProposalID)
+				}
+				if _, exists := addressesMap[key.Address]; !exists {
+					addressesMap[key.Address] = struct{}{}
+					addresses = append(addresses, key.Address)
+				}
+			}
+
+			votes, err := proposalQuery.QueryProposalVotes(proposalIDs, addresses)
+			if err != nil {
+				errors := make([]error, 0, len(keys))
+				for range keys {
+					errors = append(errors, err)
+				}
+				return nil, errors
+			}
+
+			voteKeyToVote := make(map[ProposalVoteKey]*models.ProposalVote, len(votes))
+			for _, vote := range votes {
+				key := ProposalVoteKey{vote.ProposalID, vote.VoterAddress}
+				voteKeyToVote[key] = vote
+			}
+
+			// map query results to input keys' order
+			res := make([]*models.ProposalVote, len(keys))
+			for index, key := range keys {
+				res[index] = voteKeyToVote[key]
+			}
+			return res, nil
+		},
+	})
+
+	proposalDepositLoader := godataloader.NewDataLoader(godataloader.DataLoaderConfig[ProposalDepositKey, *models.ProposalDeposit]{
+		MaxBatch: DefaultMaxBatch,
+		Wait:     DefaultWait,
+		Fetch: func(keys []ProposalDepositKey) ([]*models.ProposalDeposit, []error) {
+			// use maps to check if proposal id or address have been added to array yet
+			proposalIDsMap := make(map[int]struct{})
+			addressesMap := make(map[string]struct{})
+
+			proposalIDs := []int{}
+			addresses := []string{}
+			for _, key := range keys {
+				if _, exists := proposalIDsMap[key.ProposalID]; !exists {
+					proposalIDsMap[key.ProposalID] = struct{}{}
+					proposalIDs = append(proposalIDs, key.ProposalID)
+				}
+				if _, exists := addressesMap[key.Address]; !exists {
+					addressesMap[key.Address] = struct{}{}
+					addresses = append(addresses, key.Address)
+				}
+			}
+
+			deposits, err := proposalQuery.QueryProposalDeposits(proposalIDs, addresses)
```

Comment:
this is querying proposal ids x addresses, which could return a lot of unused results, can we make the query more precise by filtering by tuples, i.e. sql `WHERE (proposal_id, address) IN (($id1, $addr1), ($id2, $addr2))`??